firmware/hackrf_usb/sgpio_m0.s

```diff
@@ -1,108 +1,673 @@
+	cmp new_mode, #MODE_WAIT                        // if new_mode == WAIT:                 // 1
+	beq wait_loop                                   //      goto wait_loop                  // 1 thru, 3 taken
+	b idle                                          // goto idle                            // 3
+.endm
+
+.macro handle_shortfall name
+	// Handle a shortfall.
+	//
+	// Clobbers:
+	length .req r0
+	num .req r1
+	longest .req r1
+	limit .req r1
+
+	// Get current shortfall length from high register.
+	mov length, shortfall_length                    // length = shortfall_length            // 1
+
+	// Is this a new shortfall?
+	cmp length, #0                                  // if length > 0:                       // 1
+	bgt \name\()_extend_shortfall                   //      goto extend_shortfall           // 1 thru, 3 taken
+
+	// If so, increase the shortfall count.
+	ldr num, [state, #NUM_SHORTFALLS]               // num = state.num_shortfalls           // 2
+	add num, #1                                     // num += 1                             // 1
+	str num, [state, #NUM_SHORTFALLS]               // state.num_shortfalls = num           // 2
+
+	// Back up previous longest shortfall.
+	prev .req r0
+	ldr prev, [state, #LONGEST_SHORTFALL]           // prev = state.longest_shortfall       // 2
+	str prev, [state, #PREV_LONGEST_SHORTFALL]      // prev_longest_shortfall = prev        // 2
+
+\name\()_extend_shortfall:
+
+	// Extend the length of the current shortfall, and store back in high register.
+	add length, #32                                 // length += 32                         // 1
+	mov shortfall_length, length                    // shortfall_length = length            // 1
+
+	// Is this now the longest shortfall?
+	ldr longest, [state, #LONGEST_SHORTFALL]        // longest = state.longest_shortfall    // 2
+	cmp length, longest                             // if length <= longest:                // 1
+	blt \name\()_loop                               //      goto loop                       // 1 thru, 3 taken
+	str length, [state, #LONGEST_SHORTFALL]         // state.longest_shortfall = length     // 2
+
+	// Is this shortfall long enough to trigger a timeout?
+	ldr limit, [state, #SHORTFALL_LIMIT]            // limit = state.shortfall_limit        // 2
+	cmp limit, #0                                   // if limit == 0:                       // 1
+	beq \name\()_loop                               //      goto loop                       // 1 thru, 3 taken
+	cmp length, limit                               // if length < limit:                   // 1
+	blt \name\()_loop                               //      goto loop                       // 1 thru, 3 taken
+
+	// If so, reset mode to idle and return to idle loop.
+	mode .req r3
+	mov mode, #MODE_IDLE                            // mode = MODE_IDLE                     // 1
+	str mode, [state, #MODE]                        // state.mode = mode                    // 2
```

Comment:
This could clobber a new mode that was just set by the M4.

Reply:
Fixed in cb3cfe8